### PR TITLE
Optimize tensor sum

### PR DESCRIPTION
**Pull request by snehaa8**
- Addresses review comments
- Optimizes U8/I8
- Rename to tensor_sum
- Adds support for Rpp32u/Rpp32s sums other than Rpp32f
- Reuse hip memory

### CODE REVIEW

**Comment by snehaa8** [comment]
Please review the changes, Fixed QA tests too.

File: src/modules/cpu/kernel/tensor_sum.hpp [hunk 1]
```diff
@@ -28,70 +28,70 @@ RppStatus image_sum_u8_u8_host_tensor(Rpp8u *srcPtr,
...
 
-        // Image Sum without fused output-layout toggle (NCHW)
+        // Tensor Sum without fused output-layout toggle (NCHW)
         if ((srcDescPtr->c == 1) && (srcDescPtr->layout == RpptLayout::NCHW))
         {
-            alignedLength = (bufferLength / 8) * 8;
-            vectorIncrement = 8;
-            Rpp64f sum = 0.0;
-            Rpp64f sumAvx[4] = {0.0};
+            alignedLength = (bufferLength / 16) * 16;
```
Comment (r-abishek) [inline src/modules/cpu/kernel/tensor_sum.hpp:38]: @snehaa8 Pls change this wherever we have powers of 2 - like 8, 16, 32, 64.. 
https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/blob/master/src/modules/cpu/kernel/brightness.hpp#L190

File: src/modules/hip/kernel/tensor_sum.hpp [hunk 2]
```diff
@@ -0,0 +1,1439 @@
...
+
+// -------------------- Set 0 - Reduction Stage 2 --------------------
+__global__ void tensor_sum_grid_result(Rpp32u *srcPtr,
+                                       uint xBufferLength,
+                                       Rpp64u *dstPtr)
+{
+    int id_x = hipThreadIdx_x * 8;
+    int id_z = hipBlockIdx_z;
+
+    __shared__ uint partialSumShared[1024];                         // 8192 uints of src reduced to 1024 in a 1024 x 1 thread block
```
Comment (r-abishek) [inline src/modules/hip/kernel/tensor_sum.hpp:13]: Rename all instances of <>Shared to <>_smem as per new agreement for now

File: src/modules/hip/kernel/tensor_sum.hpp [hunk 3]
```diff
@@ -0,0 +1,1439 @@
...
+        int idx = hipBlockIdx_z * 4;
+        dstPtr[idx] = partialRSumShared[0];
+        dstPtr[idx + 1] = partialGSumShared[0];
+        dstPtr[idx + 2] = partialBSumShared[0];
+        dstPtr[idx + 3] = sum;
+    }
+}
+
+// -------------------- Set 1 - Reduction Stage 1 --------------------
+//Handle U8
```
Comment (r-abishek) [inline src/modules/hip/kernel/tensor_sum.hpp:379]: Space after //

File: src/modules/hip/kernel/tensor_sum.hpp [hunk 4]
```diff
@@ -0,0 +1,1439 @@
...
+            __syncthreads();
+        }
+
+        // Final store to dst
+        if (hipThreadIdx_y == 0)
+            tensorSumArr[(hipBlockIdx_z * hipGridDim_y + hipBlockIdx_y) * hipGridDim_x + hipBlockIdx_x] = partialSumSharedRowPtr[0];
+    }
+}
+
+//Handle I8
```
Comment (r-abishek) [inline src/modules/hip/kernel/tensor_sum.hpp:439]: format space - all instances

File: src/modules/hip/kernel/tensor_sum.hpp [hunk 5]
```diff
@@ -0,0 +1,1439 @@
...
+                                RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    __shared__ int partialRSumShared[16][16];                                                   // 16 rows of src, 128 reduced cols of src in a 16 x 16 thread block
+    __shared__ int partialGSumShared[16][16];
+    __shared__ int partialBSumShared[16][16];
+    int *partialRSumSharedRowPtr = &partialRSumShared[hipThreadIdx_y][0];                       // int pointer to beginning of each row in Shared
```
Comment (r-abishek) [inline src/modules/hip/kernel/tensor_sum.hpp:668]: Shared should ideally be the last word in any of our variables so we can convert to _smem

File: src/modules/hip/kernel/tensor_sum.hpp [hunk 6]
```diff
@@ -0,0 +1,1439 @@
...
+    int globalThreads_z = handle.GetBatchSize();
+    int gridDim_x = (int) ceil((float)globalThreads_x/localThreads_x);
+    int gridDim_y = (int) ceil((float)globalThreads_y/localThreads_y);
+    int gridDim_z = (int) ceil((float)globalThreads_z/localThreads_z);
+
+    if ((srcDescPtr->c == 1) && (srcDescPtr->layout == RpptLayout::NCHW))
+    {
+        Rpp32u partialSumArrLength = gridDim_x * gridDim_y * gridDim_z;
+        Rpp32u *partialSumArr;
+        hipMalloc(&partialSumArr, sizeof(Rpp32u) * partialSumArrLength);
```
Comment (r-abishek) [inline src/modules/hip/kernel/tensor_sum.hpp:1160]: Pls use the preallocated handle memory for this, and remove all hipMalloc and hipFree from here

File: utilities/test_suite/rpp_test_suite_common.h [hunk 7]
```diff
@@ -1045,8 +1044,8 @@ inline void compare_reduction_output(Rpp32f* output, string funcName, RpptDescPt
     refFile = refPath + "/../REFERENCE_OUTPUT/" + funcName + "/"+ func + ".csv";
 
     ifstream file(refFile);
-    Rpp32f *refOutput;
-    refOutput = (Rpp32f *)malloc(srcDescPtr->n * 4 * sizeof(Rpp32f));
+    Rpp64u *refOutput;
+    refOutput = (Rpp64u *)malloc(srcDescPtr->n * 4 * sizeof(Rpp64u));
```
Comment (r-abishek) [inline utilities/test_suite/rpp_test_suite_common.h:1048]: We've always used calloc for all instances, lets  do the same here to ease future searches.

File: utilities/test_suite/rpp_test_suite_common.h [hunk 8]
```diff
@@ -1087,7 +1086,7 @@ inline void compare_reduction_output(Rpp32f* output, string funcName, RpptDescPt
             matched_values = 0;
             for(int j = 0; j < 4; j++)
             {
-                int diff = abs(output[i] - refOutput[i]);
+                int diff = output[(i*4)+j] - refOutput[(i*4)+j];
```
Comment (r-abishek) [inline utilities/test_suite/rpp_test_suite_common.h:1089]: Spaces between operators

**Review by r-abishek — CHANGES_REQUESTED** [review]
Please address these

File: src/modules/hip/kernel/tensor_sum.hpp (hunk 3 above)
Comment (snehaa8) [reply src/modules/hip/kernel/tensor_sum.hpp:379]: sure

File: src/modules/hip/kernel/tensor_sum.hpp (hunk 2 above)
Comment (snehaa8) [reply src/modules/hip/kernel/tensor_sum.hpp:13]: sure

File: src/modules/hip/kernel/tensor_sum.hpp (hunk 6 above)
Comment (snehaa8) [reply src/modules/hip/kernel/tensor_sum.hpp:1160]: Introduced 
hipMemRpp32u uintReductionArr;
hipMemRpp32s intReductionArr;

in rppdefs.h and all other places and set this to size similar to maskArr(3840x2160)

**Comment by snehaa8** [comment]
Confirmed perf tests and qa tests pass.

File: src/modules/hip/handlehip.cpp [hunk 9]
```diff
@@ -245,6 +245,8 @@ struct HandleImpl
 
         hipMalloc(&(this->initHandle->mem.mgpu.rgbArr.rgbmem), sizeof(RpptRGB) * this->nBatchSize);
         hipMalloc(&(this->initHandle->mem.mgpu.maskArr.floatmem), sizeof(Rpp32f) * 8294400);    // 3840 x 2160
+        hipMalloc(&(this->initHandle->mem.mgpu.uintReductionArr.uintmem), sizeof(Rpp32u) * 8294400);
```
Comment (r-abishek) [inline src/modules/hip/handlehip.cpp:248]: Multiple 4K mem allocations aren't needed, lets just use the same maskArr, but reinterpret it as Rpp32u or Rpp32s pointer in the hip_exec function.

File: src/modules/hip/handlehip.cpp [hunk 10]
```diff
@@ -362,6 +364,8 @@ void Handle::rpp_destroy_object_gpu()
 
     hipFree(this->GetInitHandle()->mem.mgpu.rgbArr.rgbmem);
     hipFree(this->GetInitHandle()->mem.mgpu.maskArr.floatmem);
+    hipFree(this->GetInitHandle()->mem.mgpu.uintReductionArr.uintmem);
```
Comment (r-abishek) [inline src/modules/hip/handlehip.cpp:367]: Can be removed

File: src/modules/hip/handlehip.cpp (hunk 9 above)
Comment (snehaa8) [reply src/modules/hip/handlehip.cpp]: Done

**Comment by snehaa8** [comment]
Please give a glance

**Review by r-abishek — APPROVED** [review]

**Pull request merged**